### ISSUE TITLE
Failed build on Windows with -DBUILD_SHARED_LIBS=ON

### ISSUE THREAD

**Issue by RockNHawk** [facebookresearch/faiss#2909]
# Summary

I can successfuly build without  `-DBUILD_SHARED_LIBS=ON`  -> `cmake --build build --config Release --target faiss`  and get `faiss.lib` file.

But got error with `-DBUILD_SHARED_LIBS=ON` -> `cmake --build build --config Release --target faiss`

I have tried many times, here my commands & results:

# cmake -B ... This works

"C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" ^
 -B build ^
 -DFAISS_ENABLE_GPU=OFF ^
 -DPython_EXECUTABLE="D:\Application\Python\Python310\Python.exe" ^
 -DBUILD_SHARED_LIBS=ON ^
 -DBLA_VENDOR=OpenBLAS ^
 -DMKL_LIBRARIES="C:\Program Files (x86)\Intel\oneAPI\mkl\2023.1.0\lib\intel64"

-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19045.
-- The CXX compiler identification is MSVC 19.36.32532.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenMP_CXX: -openmp (found version "2.0")
-- Found OpenMP: TRUE (found version "2.0")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Found MKL: C:\Program Files (x86)\Intel\oneAPI\mkl\2023.1.0\lib\intel64
-- Found SWIG: D:/Application/swing/swigwin-4.1.1/swig.exe (found version "4.1.1") found components: python
-- Found Python: D:/Application/Python/Python310/libs/python310.lib (found version "3.10.8") found components: Development NumPy Interpreter Development.Module Development.Embed
CMake Warning (dev) at C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.26/Modules/FetchContent.cmake:1282 (message):
  The DOWNLOAD_EXTRACT_TIMESTAMP option was not given and policy CMP0135 is
  not set.  The policy's OLD behavior will be used.  When using a URL
  download, the timestamps of extracted files should preferably be that of
  the time of extraction, otherwise code that depends on the extracted
  contents might not be rebuilt if the URL changes.  The OLD behavior
  preserves the timestamps from the archive instead, but this is usually not
  what you want.  Update your project to the NEW behavior or specify the
  DOWNLOAD_EXTRACT_TIMESTAMP option with a value of true to avoid this
  robustness issue.
Call Stack (most recent call first):
  tests/CMakeLists.txt:48 (FetchContent_Declare)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- The C compiler identification is MSVC 19.36.32532.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found Python: D:\Application\Python\Python310\Python.exe (found version "3.10.8") found components: Interpreter
-- Configuring done (20.8s)

.... 
CMake Warning at faiss/CMakeLists.txt:276 (target_link_libraries):
  Target "faiss_avx2" requests linking to directory "C:\Program Files
  (x86)\Intel\oneAPI\mkl\2023.1.0\lib\intel64".  Targets may link only to
  libraries.  CMake is dropping the item.

CMake Warning at faiss/CMakeLists.txt:276 (target_link_libraries):
  Target "faiss_avx2" requests linking to directory "C:\Program Files
  (x86)\Intel\oneAPI\mkl\2023.1.0\lib\intel64".  Targets may link only to
  libraries.  CMake is dropping the item.

-- Generating done (0.4s)

# this get failed (And I have tried both `-DBLA_VENDOR=OpenBLAS` & ` -DBLA_VENDOR=Intel10_64_dyn` )
`cmake --build build --config Release --target faiss`

error summary:  `error LNK2001: Unresolved external symbol sgemm_`

I know it's about the BLA library error, I tried many times and it doesn't work. Please help me, thanks!

 
full results:

D:\faiss\>cmake --build build --config Release --target faiss
MSBuild version 17.6.3+07e294721 for .NET Framework

  Checking Build System
  Building Custom Rule D:/AI/vector/faiss/faiss-src/faiss/CMakeLists.txt
  AutoTune.cpp
  Clustering.cpp
  IVFlib.cpp
  Index.cpp
  Index2Layer.cpp
  Auto build dll exports
    正在创建库 D:/AI/vector/faiss/faiss-src/build/faiss/Release/faiss.lib 和对象 D:/AI/vector/faiss/faiss-src/build/faiss/Relea
  se/faiss.exp
random.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
ResidualQuantizer.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
LocalSearchQuantizer.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
ProductAdditiveQuantizer.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
distances.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
IndexHNSW.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
VectorTransform.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
ProductQuantizer.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
AdditiveQuantizer.obj : error LNK2001: 无法解析的外部符号 sgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
VectorTransform.obj : error LNK2019: 无法解析的外部符号 dgemm_，函数 "public: virtual void __cdecl faiss::ITQMatrix::train(__int64,
float const *)" (?train@ITQMatrix@faiss@@UEAAX_JPEBM@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj
]
LocalSearchQuantizer.obj : error LNK2001: 无法解析的外部符号 dgemm_ [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
VectorTransform.obj : error LNK2019: 无法解析的外部符号 ssyrk_，函数 "public: virtual void __cdecl faiss::PCAMatrix::train(__int64,
float const *)" (?train@PCAMatrix@faiss@@UEAAX_JPEBM@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj
]
VectorTransform.obj : error LNK2019: 无法解析的外部符号 dsyev_，函数 "void __cdecl `anonymous namespace'::eig(unsigned __int64,doub
le *,double *,int)" (?eig@?A0x1fcd164f@@YAX_KPEAN1H@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
VectorTransform.obj : error LNK2019: 无法解析的外部符号 sgesvd_，函数 "public: virtual void __cdecl faiss::OPQMatrix::train(__int64
,float const *)" (?train@OPQMatrix@faiss@@UEAAX_JPEBM@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxpro
j]
VectorTransform.obj : error LNK2019: 无法解析的外部符号 dgesvd_，函数 "public: virtual void __cdecl faiss::ITQMatrix::train(__int64
,float const *)" (?train@ITQMatrix@faiss@@UEAAX_JPEBM@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxpro
j]
ResidualQuantizer.obj : error LNK2019: 无法解析的外部符号 sgelsd_，函数 "public: float __cdecl faiss::ResidualQuantizer::retrain_AQ
_codebook(unsigned __int64,float const *)" (?retrain_AQ_codebook@ResidualQuantizer@faiss@@QEAAM_KPEBM@Z) 中引用了该符号 [D:\AI
\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
LocalSearchQuantizer.obj : error LNK2019: 无法解析的外部符号 sgetrf_，函数 "void __cdecl `anonymous namespace'::fmat_inverse(float
*,int)" (?fmat_inverse@?A0xa9c9773e@@YAXPEAMH@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
LocalSearchQuantizer.obj : error LNK2019: 无法解析的外部符号 sgetri_，函数 "void __cdecl `anonymous namespace'::fmat_inverse(float
*,int)" (?fmat_inverse@?A0xa9c9773e@@YAXPEAMH@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
LocalSearchQuantizer.obj : error LNK2019: 无法解析的外部符号 dgetrf_，函数 "void __cdecl `anonymous namespace'::dmat_inverse(double
 *,int)" (?dmat_inverse@?A0xa9c9773e@@YAXPEANH@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
LocalSearchQuantizer.obj : error LNK2019: 无法解析的外部符号 dgetri_，函数 "void __cdecl `anonymous namespace'::dmat_inverse(double
 *,int)" (?dmat_inverse@?A0xa9c9773e@@YAXPEANH@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
utils.obj : error LNK2019: 无法解析的外部符号 sgeqrf_，函数 "void __cdecl faiss::matrix_qr(int,int,float *)" (?matrix_qr@faiss@@YAX
HHPEAM@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
utils.obj : error LNK2019: 无法解析的外部符号 sorgqr_，函数 "void __cdecl faiss::matrix_qr(int,int,float *)" (?matrix_qr@faiss@@YAX
HHPEAM@Z) 中引用了该符号 [D:\AI\vector\faiss\faiss-src\build\faiss\faiss.vcxproj]
D:\AI\vector\faiss\faiss-src\build\faiss\Release\faiss.dll : fatal error LNK1120: 13 个无法解析的外部命令 [D:\AI\vector\faiss\fai
ss-src\build\faiss\faiss.vcxproj]

# Platform

OS: Windows 10

Faiss version: latest

Faiss compilation options

"C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" ^
 -B build ^
 -DFAISS_ENABLE_GPU=OFF ^
 -DPython_EXECUTABLE="D:\Application\Python\Python310\Python.exe" ^
 -DBUILD_SHARED_LIBS=ON ^
 -DBLA_VENDOR=OpenBLAS ^
 -DMKL_LIBRARIES="C:\Program Files (x86)\Intel\oneAPI\mkl\2023.1.0\lib\intel64"

cmake --build build --config Release --target faiss

Running on:
- CPU

Interface: 
- C++

# Reproduction instructions

# refer 
https://github.com/bitsun/faiss-windows/issues/5
https://github.com/facebookresearch/faiss/issues/96

**Comment by RockNHawk** [comment]
I have successfuly build with BLAS, and I truend `-DFAISS_ENABLE_C_API=ON`, but the `faiss_c.dll` export no any function, how can I fix this ?

![image](https://github.com/facebookresearch/faiss/assets/3222379/792c2b12-1820-4bb1-b343-91ff5c35894c)

the faiss.dll has export functions.
![image](https://github.com/facebookresearch/faiss/assets/3222379/ae995cc5-a335-4bbb-9a5a-d8df28bb4f39)


some info:
build with cmake
`cmake.exe --build build --config Release --target faiss_c`

**Comment by fwaris** [comment]
I believe that in windows you need '*__declspec(dllexport)*' for the function to be exported:
```c++
extern "C" void __declspec(dllexport) SomeFunction();
```
I was able to compile in windows but I don't see the exported functions either. I don't think the fix is easy.

BTW you can static link to MKL using VCPKG on windows. Here is my cmake command:
```cmd
set MKL_PATH=E:/s/repos/vcpkg/packages/intel-mkl_x64-windows/lib/intel64
cmake -Wno-dev -DFAISS_ENABLE_GPU=OFF -DFAISS_ENABLE_PYTHON=OFF -DFAISS_ENABLE_C_API=ON -DBLA_VENDOR=Intel10_64lp -DCMAKE_TOOLCHAIN_FILE=E:/s/repos/vcpkg/scripts/buildsystems/vcpkg.cmake "-DMKL_LIBRARIES=%MKL_PATH%/mkl_intel_ilp64_dll.lib;%MKL_PATH%/mkl_intel_thread_dll.lib;%MKL_PATH%/mkl_core_dll.lib" -DBUILD_SHARED_LIBS=ON -B build -S .
```
It took me a while to get this figured out (Bing Chat/ChatGPT made job much easier!).

I was trying to include faiss in a dotnet application on windows but it seems difficult. One option is to use a dotnet 'managed' C++ project that can bridge between C++ faiss and dotnet.

**Comment by RockNHawk** [comment]
@fwaris Thanks for you, I alos have tried many time. 
I was under the mistaken impression that mkl could only be used on intel cpu.
It seems you are not able to get the c api under windows either, I am using Python for .NET to call the Python version of faiss.

`

cd /d D:\AI\vector\faiss\faiss-src

set cmakeExe="C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe"
set OPENBLAS="D:\AI\vector\faiss\OpenBLAS-0.3.23-x64"
set MKL_PATH="C:\Program Files (x86)\Intel\oneAPI\mkl\2023.1.0\lib\intel64\mkl_blas95_lp64.lib"
set lapack="D:\AI\vector\faiss\references\lapack"

rem openbla build dll OK
%cmakeExe% ^
 -B build ^
 -DFAISS_ENABLE_GPU=OFF ^
 -DFAISS_ENABLE_C_API=ON ^
 -DBUILD_SHARED_LIBS=ON ^
 -DFAISS_ENABLE_PYTHON=OFF ^
 -DPython_EXECUTABLE="D:\Application\Python\Python310\Python.exe" ^
 -DBLA_VENDOR=OpenBLAS ^
 -DBLAS_LIBRARIES="%OPENBLAS%\lib\libopenblas.lib" ^
 -DLAPACK_LIBRARIES="%lapack%\lib\liblapack.lib" ^


%cmakeExe% --build build --config Release --target faiss
%cmakeExe% --build build --config Release --target faiss_c

`

**Comment by fwaris** [comment]
@RockNHawk, After more tinkering I got all of the 'index' demos to run on windows (after changing some code to use "ctime" and some other fixes).

The demo code was moved to a new VS2022 C++ project which links to faiss.lib (built from source) and MKL, statically. 

Had to switch to use latest MKL. The vcpkg did not contain *Intel10_64lp* MKL version which faiss needs.

```cmd
set MKL_PATH=C:\Program Files (x86)\Intel\oneAPI\mkl\latest\lib\intel64
cmake -Wno-dev -DFAISS_ENABLE_GPU=OFF -DFAISS_ENABLE_PYTHON=OFF -DFAISS_ENABLE_C_API=OFF -DBLA_VENDOR=Intel10_64lp "-DMKL_LIBRARIES=%MKL_PATH%/mkl_intel_lp64_dll.lib;%MKL_PATH%/mkl_intel_thread_dll.lib;%MKL_PATH%/mkl_core_dll.lib" -DBUILD_SHARED_LIBS=OFF -B build -S .
```

I will try to create C++/CLI wrapper over faiss.lib with dynamically linked MKL. If successful, I will release it as a nuget package for dotnet.

**Comment by RockNHawk** [comment]
@fwaris Cool, awesome !

I see two .NET projects working with Faiss (only linux & macOS, non-Windows library). They have some `faiss_c` C# wrappers. I hope this information can help you.

`faissmask` https://github.com/andyalm/faissmask
`vektonn-index` https://github.com/vektonn/vektonn-index

**Comment by fwaris** [comment]
FaissNet a dotnet wrapper for faiss in C++/CLI. This is a windows only version.
https://www.nuget.org/packages/FaissNet/0.5.4

**Comment by RockNHawk** [comment]
> FaissNet a dotnet wrapper for faiss in C++/CLI. This is a windows only version. https://www.nuget.org/packages/FaissNet/0.5.4

Awesome !

I see `net6.0-macos` in  nuget package version list, That means the C+/CLI can crossplatform now?

![image](https://github.com/facebookresearch/faiss/assets/3222379/0c79f237-c20c-4982-949d-ea07221372fd)

**Comment by fwaris** [comment]
I know that a pure managed dotnet assembly will run on mac and linux. 

C++/CLI is a 'mixed' assembly with native and managed parts. I don't know if a windows-compiled DLL will run as-is on mac/linux. 

And, if not, how to build such an assembly for mac/linux. This is the first time I have done something like this on windows and so know very little.

**Comment by RockNHawk** [comment]
> I know that a pure managed dotnet assembly will run on mac and linux.
> 
> C++/CLI is a 'mixed' assembly with native and managed parts. I don't know if a windows-compiled DLL will run as-is on mac/linux.
> 
> And, if not, how to build such an assembly for mac/linux. This is the first time I have done something like this on windows and so know very little.

Yeah, it's strange, because in Microsoft C++/CLI documentation, C++/CLI support for .NET Core is Windows only.

https://learn.microsoft.com/en-us/dotnet/core/porting/cpp-cli#ccli-net-core-limitations

![image](https://github.com/facebookresearch/faiss/assets/3222379/a36f268e-fb51-4f00-8922-51d3e802fd69)

**Comment by fwaris** [comment]
@RockNHawk ah, thanks for the reference. That settles its then.

I did some digging and found that the proper way to build cross-platform netcore/c++ libs is to use SWIG. Here is an article that describes the process at a high level https://iamsorush.com/posts/cpp-csharp-swig/.

I will take a look at building a SWIG based wrapper later.

**Comment by RockNHawk** [comment]
@fwaris  Cool, It use PInvoke, I know another project `CppSharp` it parse the C/C++  source code into a syntax tree, then generate the C# wrapper. `CppSharp` seems mo popular?
I switch to another vector database, because faiss does no support some feature..

**Comment by fwaris** [comment]
@RockNHawk I changed FaissNet to use pinvoke over a thin native wrapper written in c++/cmake - instead of c++/cli.

The current release is for windows only as the native wrapper is not built for other platforms, yet. However, c++/cmake is portable so support for MACOS and Linux is possible now. 

Also I changed the API so that index construction can be done with 'factory strings' as described here: https://github.com/facebookresearch/faiss/wiki/The-index-factory
